### DISCUSSION TITLE
docfx v3 - LiquidTemplate.cs - Render() - liquid file name

### DISCUSSION THREAD

**Discussion by VaclavElias** [dotnet/docfx#8152] [Q&A]
I am using Visual Studio and VS doesn't know anything about liquid by default (why would it? 🤷‍♂️🤣). 

What do you think if you wouldn't restrict the **template name** to name.html&#8203;**.liquid** but allow to name the templates just with **.html** without the **.liquid**, including the files in the _include folder. 

v3 include syntax should also allow us to use html ```{%- include main.html -%}``` so we could have the HTML syntax  in the include files as well.

The current implementation should remain but could be extended to allow html files 🤣.

_Just to clarify, the templates/includes even with **.html** file extension names still contain the liquid syntax._

Advantages:

- We would gain HTML syntax highlighting in the VS for any html file which can coexist with liquid syntax by default. Obviously the liquid syntax wouldn't be highlighted
  - ✅ this could be mitigated by @KalleOlaviNiemitalo suggestion
- We would get better readability for the HTML
  - ✅ this could be mitigated by @KalleOlaviNiemitalo suggestion
- This is similar to Jekyll and so easy to reuse in both worlds
- **New:** We could use pure HTML snippets/partials which doesn't contain any liquid syntax so this example ```{%- include main.html -%}```  would do exactly what is says, include html template (in Jekyll, it goes the same pipeline like liquid so it can still contain liquid syntax even it is .html)

https://github.com/dotnet/docfx/blob/0cdc1695a8edc09312f88d90264d4331087797eb/src/docfx/template/liquid/LiquidTemplate.cs#L41

What are your thoughts?

**Comment by paulushub** [comment]
Have you read about? [Naming rule for a Renderer file](https://dotnet.github.io/docfx/tutorial/intro_template.html#naming-rule-for-a-renderer-file)

**Reply by VaclavElias** [reply]
I did read the docs you pointed out, my suggestion above (which I updated) is about to not restrict something, which can be open for more options.

For example we have to use this syntax ```{%- include header -%}``` to include some code from _includes/header.liquid file.

I am suggesting to open it so we could use ```{%- include navigation.html -%}``` to include some code from _includes/navigation.html file.

Note: In Jekyll, this navigation.html can still contain liquid syntax

This can be easily achieved by changing

https://github.com/dotnet/docfx/blob/0cdc1695a8edc09312f88d90264d4331087797eb/src/docfx/template/liquid/LiquidTemplate.cs#L133-L134

to 

<img width="488" alt="image" src="https://user-images.githubusercontent.com/4528464/187543656-28beefce-357a-4737-954d-8fcfea8d6fde.png">

Isn't it a nicer solution to have more options?

**Updated:** Just additional comment, my suggestion allows to use also html include files, which can  by pure HTML or optionally still contain liquid syntax.

**Reply by paulushub** [reply]
> Isn't it a nicer solution to have more options?

Yes, and I understand your point.
What I meant is that they have some valid reasons for the current naming system.

For the docs.microsoft.com, they have a set of VSCode extensions for authoring the contents, 
maybe the templates are also supported - I have not tried it.

**Reply by VaclavElias** [reply]
> What I meant is that they have some valid reasons for the current naming system.

I understand. That's why I suggested extending the current naming system, so it is kept as it is but allows to accomodate my suggestions.

I started contributing to this project https://github.com/stride3d/stride and am currently looking at websites. We have got there **Jekyll** generated website and the other website for manual and docs is generated by **docfx**. You have to sync certain content in between 2 websites, including you are bringing unintentionally inconsistency, 2 different stacks, ..

It would be easier to use just one static generator and if docfx can do what Jekyll does for us (not too much in our case) then life would be much easier just using docfx. The community there is C# so docfx would be easy to understand for us.

I am looking into what changes could be made in docfx v3, so it would be easy to migrate Jekyll websites not only for us but also for others to docfx v3.

This thread discussion suggests to expand current docfx v3 functionality so Jekyll websites can be literally just copied and used as they are if the file name syntax allows html files 😊.

- [x] This topic talks about naming convention for includes and master pages. 
- [x] My other topic I will bring later is allowing to consume .html files as content like .md files are. The best part of it, that docfx has all the logic done, I have just added a new ```FileFormat.Html``` (5 files only changed) which goes through the same pipeline like ```FileFormat.Markdown``` files (https://github.com/dotnet/docfx/compare/v3...VaclavElias:jekyll?expand=1)
  - This will allow additional Jekyll content being processed by docfx without doing too much changes
- [x] The other topic might be even cooler, the ```FileFormat.Html``` could have an extra pipeline (very easy to implement) which would allow to use ```page.myVariable``` in the .html files so you could have global variables rendered not only in the master pages (which is available now) but also in the .html pages (maybe also .md, not going there right now)
  - This would again simplified Jekyll content being processed by docfx without doing too much 
- [x] The other topic, important in Jekyll is Blog. In certain .html or master pages, I need to access all pages and render e.g. first 10 titles, urls, excerpt
   - With this functionality, I think many Jekyll websites (C# and .NET community) could use just docfx

**Reply by paulushub** [reply]
> You have to sync certain content in between 2 websites, including you are bringing unintentionally inconsistency, 2 different stacks.

This is also something I am looking into. Looking at the sources of the contents for DocFX, and the unique links support, I have settled on export. That is a more general form of the process used by the PDF output currently - postprocessing the generated HTML. 
That way, DocFX will serve as intermediate processor and generate output that can be fed into other systems like the Jekyll.

However, I am currently not working with v3, but I will follow your work and progress.

**Reply by VaclavElias** [reply]
I have added another interesting feature which I am testing. This below resolve **liquid syntax** in html and md pages. That means, that any page can access global variables, not only master and partials. That is another concept which works in Jekyll. Also you can use ```{%- include myPartial -%} ```

https://github.com/VaclavElias/docfx/blob/1e9a9fff1cfafd0b7bec7474bfee57853012e87b/src/docfx/template/TemplateEngine.cs#L71-L89

**Reply by paulushub** [reply]
OK! I have to skim-through the Jekyll's Step by Step Tutorial to see where you are going (never used Jekyll) !
Powerful, but since the DocFX is multi-template based, do you plan to extend these works to mustache template in the future?

**Reply by VaclavElias** [reply]
I had to learn a bit of Jekyll because of the https://github.com/stride3d/stride project I mentioned above and I learned enough to create my own blog https://www.vaclavelias.com/ which is using GitHub Pages 😃😃.

Then I started working with docfx and realised why not to use just docfx except some very good Jekyll features are missing so that's why my exploring and this discussion. 

Moving on to the blog posts with docfx, once this is challenge is done then there is no need for Jekyll for some of us.

**Reply by VaclavElias** [reply]
Regarding mustache template, I am not familiar with this syntax so I would leave others to do it. I am going to create an example template where all my updates can be seen in action.

**Reply by paulushub** [reply]
> I am going to create an example template where all my updates can be seen in action.

That will be cool, looking forward to it.

**Reply by VaclavElias** [reply]
Here is my example and instructions in this repo https://github.com/VaclavElias/docfx-website

Just a quick summary:

- you need my docfx v3 updates
- template if you want to play with the layouts
  - you can point directly to remote repo through https:// or to a local folder like in the example website
- optionally my website to see examples
  - this is pointing to local template

**My examples website**

<img width="1122" alt="image" src="https://user-images.githubusercontent.com/4528464/188756822-3c9ea1e9-29fd-4fb9-b3f6-7e30255c9014.png">

<img width="1012" alt="image" src="https://user-images.githubusercontent.com/4528464/188756842-81291d48-091d-4cbf-b12c-4f1b34fa2af6.png">

**Reply by paulushub** [reply]
Looking good, how about Github Pages output for preview?
The page/part content seems to be largely markdown, any need for the HTML file extension?

**Reply by VaclavElias** [reply]
> Looking good, how about Github Pages output for preview?

I will try to find out..

**Edited**: It would be similar process to this https://www.jamescroft.co.uk/building-net-project-docs-with-docfx-on-github-pages/#Create_a_deployment_for_DocFX_site_to_GitHub_Pages it just need to be updated with docfx v3

> The page/part content seems to be largely markdown, any need for the HTML file extension?

The `html-page.html` example page became busy with lots of examples so it lost a purpose in this case but the point for html file extension is to have **a choice**. Some people might preference for markdown pages while others for html pages. The implementation was trivial so as I mentioned in this discussion somewhere above, there is no reason to restrict it just for markdown files if someone would like to use only html. Again, their choice if html page contains only markdown or it is 50:50 or the html is majority. 

In this case, you can just rename the file to `html-page.md` and the output is going to be the same at the moment, because html is using the same pipeline as md.

There is also another reason in jekyll to process other/any file extensions. For example, you can use sitamap.xml (any xml file) and through liquid syntax generate proper xml markup by using global variables `site.pages` and `site.posts`.

<img width="339" alt="image" src="https://user-images.githubusercontent.com/4528464/188998284-92df5bc0-c0f8-4f26-9b86-0336e15edf64.png">

Or for search.json, through liquid generate search content or any json content.

<img width="660" alt="image" src="https://user-images.githubusercontent.com/4528464/188997834-e035e606-b6eb-4ebd-914d-4151ef20b365.png">

I have also added summary to my examples above..

<img width="532" alt="image" src="https://user-images.githubusercontent.com/4528464/188997177-f4378cac-e043-4d72-8f9a-61aaf5c94429.png">

**Comment by KalleOlaviNiemitalo** [comment]
Can't you map `liquid` to the HTML editor in Visual Studio? <https://docs.microsoft.com/visualstudio/ide/reference/options-text-editor-file-extension?view=vs-2019>

**Reply by VaclavElias** [reply]
Thank you @KalleOlaviNiemitalo, I updated my original post with your suggestion which works and I use it right now.
